Tasks/AzureIoTEdgeV2/registrycredentialfactory.ts

```diff
@@ -17,21 +17,50 @@ export interface ACRRegistry {
   loginServer: string,
 }
 
+export enum RegistryEndpointType {
+  ACR,
+  Generic
+}
+
 export class RegistryCredentialFactory {
-  public static fetchACRCredential(endpointName: string, registryObject: ACRRegistry): RegistryCredential {
-    if (!endpointName || !registryObject) {
-      throw Error(`endpointName or registryName is empty when fetching ACR credential`);
+  public static fetchRegistryCredential(endpointName: string, endpointType: RegistryEndpointType): RegistryCredential {
+    if (!endpointName) {
+      throw Error("endpointName is empty when fetching docker credential");
+    }
+
+    switch (endpointType) {
+      case RegistryEndpointType.ACR: {
+        let acrObject: ACRRegistry = JSON.parse(tl.getInput("azureContainerRegistry"));
+        if (!acrObject) {
+          throw Error("No Azure Container Registry configured in task.")
+        }
+        return this.fetchACRCredential(endpointName, acrObject.loginServer);
+      }
+      case RegistryEndpointType.Generic: {
+        return this.fetchGenericCredential(endpointName);
+      }
+      default: {
+        throw Error(`Endpoint type ${endpointType} is not supported yet.`)
+      }
     }
-    return new RegistryCredential(tl.getEndpointAuthorizationParameter(endpointName, 'serviceprincipalid', true),
-      tl.getEndpointAuthorizationParameter(endpointName, 'serviceprincipalkey', true),
-      registryObject.loginServer);
   }
 
-  public static fetchGenericCredential(endpointName: string): RegistryCredential {
-    if (!endpointName) {
-      throw Error(`endpointName or registryName is empty when fetching Generic credential`);
+  private static fetchACRCredential(endpointName: string, registryLoginServer: string): RegistryCredential {
+    return new RegistryCredential(
+      tl.getEndpointAuthorizationParameter(endpointName, "serviceprincipalid", true),
+      tl.getEndpointAuthorizationParameter(endpointName, "serviceprincipalkey", true),
+      registryLoginServer);
+  }
+
+  private static fetchGenericCredential(endpointName: string): RegistryCredential {
```

Comment:
the check here seems duplicate with fetchRegistryCredential. Is it possible acr will go into this method?

Reply:
Yes, acr may go into this method. This function checks different thing than `fetchRegistryCredential`.

`fetchRegistryCredential` checks the container registry type in our task configuration:
![image](https://user-images.githubusercontent.com/16605901/75509583-2176a400-5a23-11ea-8921-e4c8c524ad82.png)

And this function (`fetchGenericCredential`) checks the container registry type in a service connection, which belongs to Generic Container Registry above:
![image](https://user-images.githubusercontent.com/16605901/75510080-ac0bd300-5a24-11ea-9ba1-f4d378450b03.png)

So if a user chooses `Generic Container Registry`, and then selected a service connection which configures an ACR, the ACR part in this function will be invoked.